SavedModel export script for Music VAE

quantumpencil opened the pull request:
Script exports a saved model.pb with 3 signature defs, sample/encode/decode based on the interface in trained_model.py. Not sure if sample signature def is necessary as it is basically the same thing as decode, except z is not optional. 

TrainedModel's decode optionally returns decoder_results, but this is not a tensor so I'm confused as to what this is/if it is used. Will remove that comment after clarification

Upcoming PR will include a ServableVAEModel class which can be imported to query an instance of this savedmodel running in TF serving

quantumpencil commented:
@googlebot I signed it!

adarob commented on magenta/models/music_vae/music_vae_export.py:
```diff
@@ -0,0 +1,219 @@
+from __future__ import division
+from __future__ import print_function
+
+import argparse
```
please use absl.flags

quantumpencil replied:
Done

adarob commented on magenta/models/music_vae/music_vae_export.py:
```diff
@@ -0,0 +1,219 @@
+
+
+# TODO create a named registry of signature def builders keyed by name for better type safety
+# Sample seems almost exactly like decode from the trained model code, possibly redundant
```
The difference is that sample feeds in a sample from a random normal distribution in place of the Z

quantumpencil replied:
Cool -- I decided to just remove the signature as it's basically a matter of arguments, and the ServableModel class coming up will have handle passing an appropriately sampled random vector passed in to decode

adarob commented on magenta/models/music_vae/music_vae_export.py:
```diff
@@ -0,0 +1,219 @@
+    }
+    if input_tensors.c_input is not None:
+        sample_inputs['c_input'] = saved_model.utils.build_tensor_info(input_tensors.c_input)
+    if input_tensors.z_input is not None:
```
as mentioned above, you should feed in a random normal sample for the z

quantumpencil replied:
See above

adarob commented on magenta/models/music_vae/music_vae_export.py:
```diff
@@ -0,0 +1,219 @@
+def build_encode_signature(input_tensors, output_tensors):
+    # define the signature for the encoding operation
+    encode_inputs = {
+        'inputs': saved_model.utils.build_tensor_info(input_tensors.inputs),
```
use either ' or " throughout

adarob commented on magenta/models/music_vae/music_vae_export.py:
```diff
@@ -0,0 +1,219 @@
+    if input_tensors.c_input:
+        decode_inputs["c_input"] = saved_model.utils.build_tensor_info(input_tensors.c_input)
+
+    # Question for magenta team about this, appears in trained model as a return value from decode,
```
These are really just used internally for the hierarchical decoder to be able to pass state around.

quantumpencil replied:
Great, removed

quantumpencil commented:
@adarob Thank you so much, I'll address these issues. I realized today that the CLA I signed actually needs to be a corporate one -- what process should I go through to void my individual CLA and sign the corporate one?

adarob commented:
You should be able to follow the instructions here: https://cla.developers.google.com/about/google-corporate

Let me know if you have issues.

quantumpencil commented:
Great, I just heard that the corporate CLA has been cleared/signed, so we're good to go.

quantumpencil commented:
@adarob Sorry about the delay, had to push to cut an internal release. I've addressed your comments and made a couple small changes to config (a few new fields, and a method for serializing information needed in order to properly serve routines trained with the existing training infrastructure as part of the script)

The ServableModel/Servable VAE module I'll be PRing in uses the exported metadata when constructing the a new interface to the actual saved model running on tf-serving (simplified implementations of sample/interpolate in terms of the 'encode' and 'decode' signatures exported here, via PredictRequest proto over grpc) The new fields are strictly additive and default initialized so no existing code is affected.

Let me know if you have any other concerns.
